src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs

```diff
@@ -63,6 +63,28 @@ public override void Contextualize(ViewContext viewContext)
             }
 
             ViewData = viewContext.ViewData as ViewDataDictionary<TModel>;
+
+            if (ViewData == null)
+            {
+                // The view data that we have at this point might be of a more derived type than the one defined at compile time.
+                // For example ViewDataDictionary<Cat> where our TModel is animal and Cat : Animal.
+                // This can happen if someone modified the model type through the page application model.
+                // In that case, we check if the type of the current view data, 'ViewData<TRuntime>' is "covariant" with the
+                // one defined at compile time 'ViewData<TModel>'
+                // When a type T<K> is covariant it means that an instance of that type T<'a> can be safely assigned to a variable
+                // of type T<'b> when 'a is of a more derived type of 'b. An example of this would be the array type, in which an
+                // instance of an array int[] can be safely assigned to a variable of type object[]. Or in code,
+                // var int [] integers = new[]{ 1, 2, 3, 4};
+                // var object [] data = integers;
+                // The statement above would not be possible in the general case given the fact that int[] doesn't extend object[], but
+                // its valid because arrays are covariant and int inherits from object.
+                var runtimeType = GetViewDataOfTRuntimeType(viewContext);
```

Comment:
This helper is doing scary runtime Reflection. But, it should just return `viewContext.ViewData.ModelMetadata.ModelType`. Use that instead and confirm your tests still work.